optional try flattening questions

bjhomer opened the pull request:
This PR just serves as a place to ask some questions about the SILGen portion of `try?` flattening implementation. It will not be merged in its current state.

bjhomer committed:
Make 'try?' flatten optional chaining like optional-chaining does.

If the sub-expression of the 'try?' is optional, the result will be the same level of optional-ness.
If the sub-expression is non-optional, the result is optional.

Thus, the following lines all end up with the same type of 'Int?'
 - let x = try? 3 as Int
 - let x = try? 3 as? Int
 - let x = try? 3 as Int?

bjhomer commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1520,16 +1520,7 @@ RValue RValueEmitter::visitOptionalTryExpr(OptionalTryExpr *E, SGFContext C) {
-    ManagedValue wrapped = SGF.getOptionalSomeValue(E, subExprValue, optTL);
-    branchArg = wrapped.forward(SGF);
-  }
+  SGF.emitExprInto(E->getSubExpr(), optInit);
```
I'm no longer managing the `some`/`none` directly here; instead, just taking the value from the sub-expression. I still need to emit it into an existing variable, though, so I can modify the result if the expression throws.

Am I doing this right? Is it okay to not treat `isByAdddress` any differently?

rjmccall replied:
No, you still need to do the two paths here.  But you no longer need to worry about the injection, so you can just call `SGF.emitExprInto(E->getSubExpr(), optInit);` in the `isByAddress` path and then `ManagedValue wrapped = SGF.emitRValueAsSingleValue(E->getSubExpr()); branchArg = wrapped.forward(SGF);` in the scalar path.

bjhomer commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1555,10 +1547,7 @@ RValue RValueEmitter::visitOptionalTryExpr(OptionalTryExpr *E, SGFContext C) {
-    SGF.B.createBranch(E, contBB);
-  else
-    SGF.B.createBranch(E, contBB, branchArg);
+  SGF.B.createBranch(E, contBB);
```
Since `branchArg` no longer exists, I'm assuming that these can be collapsed to a unified case.

rjmccall commented:
Did that work?

bjhomer commented:
I believe so. I was getting stuck on an assertion due to another bug. (I was using `expr->getType()` instead of `cs.getType(expr)` in `CSApply.cpp`) I just got that worked out about 30 minutes ago, and in my initial tests everything is now working. I still feel like I'm coding blind on the SILGen stuff, so I'll certainly hope for some scrutiny there.

I still need to run through the rest of the test suite and fix things up as needed.

bjhomer commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1590,8 +1587,6 @@ RValue RValueEmitter::visitOptionalTryExpr(OptionalTryExpr *E, SGFContext C) {
   }
 
-  optInit->finishInitialization(SGF);
-
```
Now that we're using `emitExprInto` above, it already calls `finishInitialization`. Calling it again here causes an assert.

(Note that `emitExprInto` always happens on the `isByAddress == true` branch, and this `finishInitialization` call also only happens when `isByAddress == true` due to the `return` in the `if` block above.)

rjmccall replied:
This was only necessary before because of the split optional/value initialization, so it's right to remove it.

bjhomer commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1571,8 +1570,6 @@ RValue RValueEmitter::visitOptionalTryExpr(OptionalTryExpr *E, SGFContext C) {
 
   if (isByAddress) {
-    SGF.emitInjectOptionalNothingInto(E,
-                    optInit->getAddressForInPlaceInitialization(SGF, E), optTL);
```
I'm unsure about whether this is needed or not. I believe this portion is building the `catch` portion of the try handling, so it seems like it we probably need something to make sure that the result is set to `nil` in the case that an error was thrown.

But I'm also concerned that `optInit` has already had `finishInitialization()` called from within `emitExprInto()` up above. Does this indicate that I shouldn't be using `emitExprInto` above? Or is it okay to call this line after initialization is finished?

rjmccall replied:
You're right to be concerned.  There currently isn't any way to re-use an Initialization like this in different control-flow paths, so you need to pattern the code on what emitOptionalEvaluation does, where it creates a new Initialization (with `useBufferAsTemporary`) to emit the expression into, then cancels its cleanup and finishes the original Initialization.

...if you're feeling adventurous, you could try to add this sort of branching support as a feature to Initialization; I think all of our `Initialization` implementations should allow it.

bjhomer committed:
Adjust SILGen tests for flattened optional try.

bjhomer committed:
Add backwards-compatible behavior for swift-version < 5

bjhomer committed:
Add a migration for the Swift 5 'try?' changes.

This migrates anything using the following pattern:

```swift
if let optX = try? someOptional(),
    let x = optX,
    ...
```

to this:

```swift
if let x = try? someOptional(),
    ...
```

Commit:
[stdlib] Update Policy.swift to use designated protocols for operators.

Also enable the parsing of these with
-enable-operator-designated-protocols in the stdlib build.

The constraint solver support to make use of these is not yet in-tree.
and will probably land with a different command-line option to enable
it, so that we can continue to have parsing support enabled in-tree
while the constraint solver support is iteratively improved.

drodriguez committed:
[android] adb_push_build_products.py support for aarch64.

A new parameter for the adb_push_build_product.py script to allow
pushing the right libc++_shared.so for the host architecture.

mikeash committed:
[Testing] Add a substitution for %target-is-simulator.

mikeash committed:
[IRGen][Runtime] Adjust the ObjC reserved bits on x86-64 to exactly match what the target uses.

Previously we had a single mask for all x86-64 targets which included both the top and bottom bits. This accommodated simulators, which use the top bit, while macOS uses the bottom bit, but reserved one bit more than necessary on each. This change breaks out x86-64 simulators from non-simulators and reserves only the one bit used on each.

rdar://problem/34805348 rdar://problem/29765919

jrose-apple committed:
Allow '@usableFromInline' on 'dynamic' decls

It actually /does/ make sense to enforce the usable-from-inline rules
on dynamic declarations, but that would break source compatibility at
this point. Just allow '@usableFromInline' to be written on 'dynamic'
declarations, so that it'll be compatible with Swift 5.

jrose-apple committed:
Remove unwanted exceptions from @inlinable checking

- Treat protocol requirements just like associated types in how they
  inherit usable-from-inline-ness, rather than special-casing them in
  the check for inlinable code.

- Enum elements are already treated this way, so we can remove a
  redundant check for that.

- Finally, start enforcing that 'dynamic' declarations need to be
  '@usableFromInline' to be used in inlinable functions...in Swift 5
  mode or resilient code. I didn't even add a warning in Swift 4/4.2
  because it was illegal to use '@usableFromInline' on a 'dynamic'
  declaration in Swift 4.2. (Oops.)

mikeash committed:
[Stdlib] Change SWIFT_RUNTIME_STDLIB_INTERNAL to not export the symbol.

The functions in LibcShims are used externally, some directly and some through @inlineable functions. These are changed to SWIFT_RUNTIME_STDLIB_SPI to better match their actual usage. Their names are also changed to add "_swift" to the front to match our naming conventions.

Three functions from SwiftObject.mm are changed to SPI and get a _swift prefix.

A few other support functions are also changed to SPI. They already had a prefix and look like they were meant to be SPI anyway. It was just hard to notice any mixup when they were #defined to the same thing.

rdar://problem/35863717

slavapestov committed:
AST: Use SubstitutionMap::getReplacementTypesBuffer() instead of getReplacementTypes() in a couple of places

lorentey committed:
[stdlib] Set, Dictionary: Flatten switch statements

The optimizer dislikes nested switch statements; flatten them out to simplify optimization and to hopefully speed things up a little.

Commit:
[SourceKit] Make each Variant keep a strong reference to its SourceKitdResponse context

sourcekitd_variant_t is only safe to use while the sourcekitd_response_t it was
retrieved from is still alive, so keep a strong reference to SourceKitdResponse
(the Swift wrapper of sourcekitd_response_t) in each Variant (the Swift wrapper
of sourcekitd_variant_t).

mikeash committed:
[Stdlib] Pull _swift_stdlib_random into a separate file.

mikeash committed:
[Stdlib] Move the shims used by Platform.swift into a separate header and make them static inline. This avoids exposing them as exported symbols in libswiftCore.

mikeash committed:
[Stdlib] Move _swift_stdlib_read/write/close into LibcShimsInline.h.

lorentey committed:
[stdlib] Bitset: Use explicit left/right shifts instead of multiplication/division by powers-of-two

The divisions don’t get eliminated in -Osize builds.

lorentey committed:
[stdlib] Dictionary: Force-inline native iterators

compnerd committed:
SourceKit: explicitly link against BlocksRuntime as well

On Darwin platforms, libdispatch and libBlocksRuntime are re-exported from
libSystem (via LC_REEXPORT_DYLIB).  Other platforms do not have libdispatch and
libBlocksRuntime in their C runtime, so we need to explicitly link against them.
Now that we are building BlocksRuntime with hidden visibility, we do not
accidentally get the symbols from libdispatch.

lorentey committed:
[stdlib] Dictionary, Set: Force-inline hidden Sequence/Collection customization points

This should eliminate a branch, which should probably lead to a tiny overall code size improvement, as well as a tiny performance boost.

eeckstein committed:
COWArrayOpts: remove the -view-cfg-before-cow-for option for debugging

This option is very old. The same effect can now be achieved with pass-manager options, like -sil-print-before

lorentey committed:
[stdlib] Force-inline trivial default implementations for hidden Sequence/Collection customization points

This is usually a code size pessimization, but in this case the bodies are trivial, so inlining them eliminates a call + a conditional branch.

lorentey committed:
[stdlib] _Bitset: Revert to using division/multiplication

However, convert input values to UInts before dividing them.

lorentey committed:
[stdlib] _Bitset: Use &* for join and capacity

We know it won’t overflow, because the result will not exceed the capacity, which is guaranteed to fit in an integer.

mikeash committed:
[Stdlib] Make all the functions in LibcShims.h either INTERNAL or inline. Move LibcShimsInline.h to LibcOverlayShims.h for more consistent naming. Fix up several tests that needed the mock Darwin overlay built. Fix one SourceKit test that no longer produces is_system: 1 on an `import Darwin` line.

compnerd committed:
SourceKit: use shared BlocksRuntime

Adjust the SourceKit's embedded libdispatch build to use the shared build of
BlocksRuntime now that we can generate that.

gottesmm committed:
[silgen] Change silgen pattern to use destructure_tuple instead of borrow + tuple_extract + copy.

rdar://29791263

lorentey committed:
Merge pull request #19664 from lorentey/dictionary-subscript-setter-is-slow

[stdlib] Speed up _Bitset in size-optimized builds

lorentey committed:
[stdlib] Force-inline native Set iterators

mikeash committed:
[Stdlib] Change _swift_stdlib_random SPI to swift_stdlib_random API.

rudkx committed:
Merge pull request #19677 from rudkx/join-protocol-compositions

[TypeJoin] Implement Type::join for protocols and protocol compositions.

lorentey committed:
Merge pull request #19684 from lorentey/force-inline-native-iterators

[stdlib] Set, Dictionary: force-inline native iterators

lorentey committed:
Merge pull request #19683 from lorentey/inline-customization-points

[stdlib] Force-inline some Sequence/Collection customization points

mikeash committed:
[Stdlib] Move the declaration of __swift_ssize_t to SwiftStddef.h so Random.h/cpp can use it.

mikeash committed:
[Stdlib] Clean up includes in LibcShims and Random.

mikeash committed:
[Stdlib] Delete a stray ß in LibcShims.h

lorentey committed:
[stdlib] Set, Dictionary: Switch to using a _BridgeObject instead of an enum

lorentey committed:
[stdlib] Set, Dictionary: Implement modify accessor for _variant.asNative

lorentey committed:
[stdlib] Fix building on platforms without an Objective-C runtime

lorentey committed:
[stdlib] Dictionary.Values.swapAt: Fix uniqueness check

lorentey committed:
[stdlib] Rework Dictionary.values._modify

mikeash committed:
[Stdlib] Fix Linux detection for malloc_size. Silence const cast warnings. Fix Linux implementation in Random.cpp to actually compile.

lorentey committed:
[test] Update Set/Dictionary tests

Commit:
[ConstraintSystem] Add a staging option for the operator protocols work.

Add a staging option for enabling upcoming changes in the solver in
support of using the protocols designated in Policy.swift to guide
operator type checking.

airspeedswift committed:
Underscore @usableFromInlinable symbols (#19686)

lorentey committed:
[test] Set and Dictionary now has an extra inhabitant

lorentey committed:
[stdlib] Set._Variant: Add a dummy initializer for use in _modify

Dictionary already does this.

lorentey committed:
[stdlib] Allow native dictionaries to advance Cocoa indices

lorentey committed:
[stdlib] Set, Dictionary: Reorganize equality checks to fix Linux builds

xedin committed:
[ConstraintSystem] Form argument constraints instead of using `matchTypes` directly

Forming constraints instead of using `matchTypes` while matching argument/parameter
types makes sure that `failedConstraint` points to failed argument conversion
instead of `applicable function` constraint, which is better for diagnostics.

compnerd committed:
SourceKit: correct header include order

Internal.h must be included before sourcekitd.h to ensure that we get the
correct definition for Windows targets.  The inclusion of `sourcekitd.h` is not
needed after the inclusion of `Internal.h` so simply remove the inclusion.

nkcsgexi committed:
swift-module-digester: use getEffectiveAccess() to get accessibility when checking ABI stability.

This allows us to include internal decls with @usableFromInline attribute, whose stored property
changes can effect ABI.

nkcsgexi committed:
[test] regenerated ABI checker baseline to include decls with @usableFromInline.

Commit:
[ConstraintSystem] Use operator designated protocols to speed type checking.

Use the type designated in the operator declaration as a mechanism to
limit the exploration we do in constraint solving. If we can find a
solution with an overload defined within that type or an extension of
that type, we do not look at any of the other options in the
disjunction.

This is disabled by default, and enabled with
> `-swift-version 5 -Xfrontend -solver-enable-operator-designated-protocols`

lorentey committed:
[stdlib] Fix build on 32-bit platforms

nkcsgexi committed:
swift-module-digester: include underscored and unavailable decls when checking stdlib abi stability.

jrose-apple committed:
[Frontend] Allow -emit-interface with -typecheck (#19676)

Textual module interfaces don't actually depend on SILGen, so we
shouldn't need to run SILGen (or serialize an entire binary module) if
we're just trying to emit a textual interface. On the other hand, if
we /are/ going to run SILGen and then SIL diagnostics, we shouldn't
delay those diagnostics by spending time emitting a textual interface,
or for that matter a TBD file.

Using this, update all the ModuleInterface tests that use
`-emit-module -o /dev/null` to use `-typecheck` instead, except for
those using `-merge-modules`.

harlanhaskins committed:
[stdlib] Break infinite recursion false positive in _fatalErrorMessage (#19693)

Introduce an inner function that breaks the infinite recursion check, silencing this warning:

```
swift/stdlib/public/core/AssertCommon.swift:166:15: warning: all paths through this function will call itself
internal func _fatalErrorMessage(
              ^
```

nkcsgexi committed:
Merge pull request #19697 from nkcsgexi/effective-access

swift-module-digester: use getEffectiveAccess() to get accessibility when checking ABI stability.

Commit:
Test updates for -solver-enable-operator-designated-protocols.

These are cases that I know are faster when this is enabled. The test
updates all additionally disable the existing performance hacks as
well as the shrink phase of the solver.

Commit:
Merge pull request #19689 from nathawes/keep-sourcekitd-response-alive-while-variant-lives

[SourceKit] Make each Variant keep a strong reference to its SourceKitdResponse context

drodriguez committed:
build: Fix Android 64 bit architecture name.

swift-ci committed:
Merge pull request #19680 from compnerd/exports

Commit:
Fix release-build test failure due to unaccounted-for warning.

Commit:
Fix small formatting problem.

Commit:
[ConstraintSystem] Only look for overloads in the designated type if
one was designated.

jrose-apple committed:
[SDK] Get the preferred @encode string for MapKit types dynamically (#19691)

On older OSs, CLLocationCoordinate2D was an anonymous struct with a
typedef, which meant its Objective-C type encoding string was "{?=dd}"
instead of "{CLLocationCoordinate2D=dd}". This incompatibility led to
us rejecting casts from NSValues created with CLLocationCoordinate2Ds
in Objective-C on older OSs because they didn't match the type
encoding provided. Instead, we can try to create an NSValue the way
the frameworks do, and see what /its/ type encoding is. This is what
SceneKit already does.

There's an extra wrinkle here because the convenience methods for
encoding CLLocationCoordinate2Ds are actually added in MapKit rather
than CoreLocation. That means that if someone's app just uses
CoreLocation, we can't rely on those convenience methods to get the
correct type encoding. In this case, the best thing we can do is just
give up and use the static encoding.

This whole thing is indicative of a bigger problem, namely that
NSValue normally doesn't try to validate types at all. However, Swift
bridge casting really does want to distinguish, say, a
CLLocationCoordinate2D from a CGPoint, and checking the NSValue encode
string was a way to do that. But this shows that it's still not safe
to assume that's consistent against everything in a process even if
they're all using @encode on the real struct (or the Swift
equivalent), because the different parts of the process may have been
compiled against different SDKs.

This change does not attempt to solve that problem.

Finishes rdar://problem/44866579

airspeedswift committed:
Additional ownership annotations

slavapestov committed:
IRGen: Don't emit empty module records in YAML type info dump

slavapestov committed:
IRGen: Add a new 'Legacy' lowering mode that loads type info from a YAML file

The YAML format is the same one produced by the -dump-type-info
frontend mode.

For now this is only enabled if the -read-type-info-path frontend
flag is specified.

Progress on <rdar://problem/17528739>.

Commit:
[ConstraintSystem] Replace argument walk with a more robust utility.

For now this is only used for the same purpose as the previous type
constraint walking utility function.

Eventually we can incorporate this into our selection of overload
choices.

compnerd committed:
Merge pull request #19668 from compnerd/explicit-BlocksRuntime

SourceKit: explicitly link against BlocksRuntime as well

compnerd committed:
build: be a bit more resilient

In the case that a system update temporarily breaks the command, the output may
be empty, and this can cause CMake to fail in a weird manner.  Quote the
variable to be more resilient to the transient failure.

compnerd committed:
SourceKit: enable builds for Windows

SourceKit's build is now ammenable to building for Windows.  Generalize the path
to enable building it for Windows as well as Linux.  The libdispatch build for
the compiler is needed for all non-Darwin targets currently.

gottesmm committed:
Merge pull request #19700 from gottesmm/pr-ef927a3d187515aa9a630c73d9aeff2bbde00aa5

[silgen] Change silgen pattern to use destructure_tuple instead of bo…

Commit:
[ConstraintSystem] Distinguish between overloads from a type vs. an extension.

For operators with default implementations in an extension, we don't
want to typecheck it both with overloads from the protocol type and
the ones from the extension.

slavapestov committed:
IRGen: Factor out LegacyLayoutFormat.h and add a descriptive comment

slavapestov committed:
IRGen: Remove some unused and used-once functions

A few utility methods would bypass computing the TypeInfo for a tuple,
but they were only used in assertions or used in places where I can't
imagine tuples coming up often enough for it to matter.

swift-ci committed:
Merge pull request #19701 from drodriguez/fix-android-archs

slavapestov committed:
IRGen: We no longer deallocate partially-initialized foreign classes

This was an old workaround for a bug that should no longer
happen now that convenience initializers only have allocating
entry points.

DougGregor committed:
[ABI] Only emit resilient superclass descriptor references in class metadata.

The superclass descriptor reference in class context descriptors is only used
for metadata bound computations when the superclass is resilient. Only
include the superclass descriptor reference when the class has a resilient
superclass, using a trailing record. It’s a tiny space savings for
classes that don’t have resilient superclasses.

swift-ci committed:
Merge pull request #19704 from compnerd/quoting

DougGregor committed:
[ABI] Add a mangled superclass type to class context descriptors.

This allows us to recover the complete superclass type from metadata.
Currently unused, but we want to stub out the space in the ABI.

DougGregor committed:
[Runtime] Check demangled superclass names against compiler-provided super.

Introduce a sanity check verifying that we can demangle the superclass
of a class when forming type metadata, and that the result matches the
compiler-provided superclass metadata.

slavapestov committed:
IRGen: TypeConverter::convertType() always returns a complete type

nikiltabraham committed:
Update README.md

Commit:
[ConstraintSystem] Address review feedback.

rintaro committed:
[CodeCompletion] Fix completion for local decls in trailing closure

Previously, local decls in trailing closure didn't show up if the
closure had preceding arguments and the completion was triggered at
beginning position of expression context. like:

   funcName(x: arg1) {
      var localVar = 12
      if <HERE>
   }

The completion mode used to be overwritten in 'completeCallArg()' which
is called from 'parseExprCallSuffix(). We should detect completion for
immediate argument position in 'parseExprList()'.

rdar://problem/41869885

lorentey committed:
Merge pull request #19694 from lorentey/values-isnt-really-a-mutablecollection

[stdlib] Allow native dictionaries to advance Cocoa indices

lorentey committed:
[stdlib] _Bitset: Reduce unnecessary overflow checking

This is mostly a code size improvement.

lorentey committed:
[stdlib] _SetBuilder, _DictionaryBuilder: Assume target has enough capacity

The compiled code included unnecessary references to resize().

lorentey committed:
[stdlib] Eliminate more overflow checks

This is a tiny code size improvement.

mikeash committed:
[Stdlib] Make _swift_extractDynamicValue INTERNAL as it's used by Foundation.

airspeedswift committed:
reorder a __consuming func

lorentey committed:
[stdlib] Acknowledge there are no spare bits in Set/Dictionary’s BridgeObject storage

We really need to do a pass on the _BridgeStorage API at some point soon.

rudkx committed:
Merge pull request #19698 from rudkx/designated-protocols

[ConstraintSystem] Use operator designated protocols to speed type checking.

benlangmuir committed:
Merge pull request #19709 from rintaro/ide-completion-trailingclosurelocal-rdar41869885

[CodeCompletion] Fix completion for local decls in trailing closure

lorentey committed:
[stdlib] Dictionary.Index: Remove duplicate definition for _isNative

xedin committed:
Merge pull request #19699 from xedin/simplify-applicable-fn-to-create-constraints-for-args

[ConstraintSystem] Form argument constraints instead of using `matchT…

swift-ci committed:
Merge pull request #19708 from nikiltabraham/patch-1

Commit:
[ConstraintSystem] Tweak -disable-constraint-solver-performance-hacks.

Allow a couple of the tests in shortCircuitDisjunctionAt() to run even
when hacks are disabled.

The first of these tests is imperfect since it assumes that "favored"
disjunction choices all come before the others (which may be the case
now but is pretty brittle).

Similarly, the second of these tests assumes that choices that are
fixes always come after ones that are not.

What we really want to do is skip these choices rather than stopping
at these points.

Commit:
Revert "[SourceKit] Make each Variant keep a strong reference to its SourceKitdResponse context"

johnno1962 committed:
Fix problem preventing new build system from using custom toolchains (#19509)

eeckstein committed:
COWArrayOpts: make the optimization work again for two-dimensional arrays.

With removing of pinning and with addressors, the pattern matching did not work anymore.
The good thing is that the SIL is now much simpler and we can handle the 2D case without pattern matching at all.
This removes a lot of code from COWArrayOpts.

rdar://problem/43863081

Commit:
Merge pull request #19712 from apple/revert-19689-keep-sourcekitd-response-alive-while-variant-lives

Revert "[SourceKit] Make each Variant keep a strong reference to its SourceKitdResponse context"

swift-ci committed:
Merge pull request #19706 from airspeedswift/further-pwnage

nkcsgexi committed:
swift-module-digester: diagnose self access kind changes for function decls.

eeckstein committed:
swift-demangle: Also accept $s symbols without the $ prefix.

This makes it easier to copy-paste $s symbols with a double-click.
We already had this for $S, now it also works for the new mangling prefix $s.

mikeash committed:
[Stdlib] Make the !_runtime(_ObjC) version of _bridgeAnythingToObjectiveC non-inlinable as it calls an internal function.

swift-ci committed:
Merge pull request #19715 from eeckstein/improve-swift-demangle

mikeash committed:
Merge pull request #19540 from mikeash/reserved-bits-refinement

[IRGen][Runtime] Adjust the ObjC reserved bits on x86-64 to exactly match what the target uses.

compnerd committed:
test: simplify the checks (NFC)

Use pointer size to determine 32-bit vs 64-bit and then check appropriately.
Use a shared check to ensure that the same pattern occurs on all targets.  NFC.

Commit:
Move another typechecker performance test to run with -solver-enable-operator-designated-protocols.

harlanhaskins committed:
[SILOptimizer] Clean up infinite recursion diagnostic pass (#19710)

* [SILOptimizer] Clean up infinite recursion diagnostic pass

NFC. This cleans up the pass to avoid the extra lambda and to clear up
the ordering of what check is supposed to come before what.

* Address review comments

compnerd committed:
Merge pull request #19674 from compnerd/shared-blocks-are-back

SourceKit: use shared BlocksRuntime

DougGregor committed:
[ABI] Use mangled superclass names from class context descriptors.

Rather than rely on the metadata initialization function to compute and
fill in the superclass, use the mangled superclass name to construct the
superclass metadata.

DougGregor committed:
[IRGen] Ensure that we emit metadata for types referenced from mangled names.

nkcsgexi committed:
swift-module-digester: include parameters' value ownership values in the digests.

Commit:
Revert a type checker perf test that apparently is still not consistently fast enough.

mzp committed:
[SR-8340]Improve fix-it for var and subscript in Protocol (#19660)

* [Parser] Improve fix-it for subscription in protocol
* [Sema] Add fix-it for property in protocol

https://bugs.swift.org/browse/SR-8340

nkcsgexi committed:
[test] Regenerate ABI/API checker baseline by using stdlib without assertion enabled.

If we use a baseline generated from assertion build, the checker will report false-positives
when running in a non-assertion build since some decls only exist in the former.

resolves: rdar://45014723

compnerd committed:
Merge pull request #19718 from compnerd/simplify-test

test: simplify the checks (NFC)

slavapestov committed:
IRGen: Type lowering cache doesn't need to store 'forward declarations'

swift-ci committed:
Merge pull request #19720 from rudkx/revert-test-change

nkcsgexi committed:
Merge pull request #19721 from nkcsgexi/valueownership

[test] Regenerate ABI/API checker baseline by using stdlib without assertion enabled.

adrian-prantl committed:
Debug Info: Normalize paths when creating DIFile objects.

For a yet-to-be-determined reason the Clang SourceManager returns
source file paths with a ./ component before the filename if the
module cache is cold.

<rdar://problem/43439465>

jrose-apple committed:
[Driver] Expose -emit-parseable-module-interface[-path]

Commit to a command line option spelling so that build systems can
start testing it. I deliberately picked one of the longer names we
were considering because we can always decide to add a shorter alias,
but can't decide a shorter name was too generic.

Like the other supplementary output flags,
-emit-parseable-module-interface-path will emit a .swiftinterface file
to a particular path, while -emit-parseable-module-interface will put
it next to the main output (the one specified with -o).

rdar://problem/43776945

jrose-apple committed:
Rename "textual interface" to "parseable interface" (#19713)

We already have something called "module interfaces" -- it's the
generated interface view that you can see in Xcode, the interface
that's meant for developers using a library. Of course, that's also a
textual format. To reduce confusion, rename the new module stability
feature to "parseable [module] interfaces".

jrose-apple committed:
[Driver] Disallow emitting swiftinterfaces with bridging headers

There's no place to put the bridging header in a swiftinterface, and
they don't make sense with the intended use case of distribution.
Just disallow it up front.

rdar://problem/44113493

nkcsgexi committed:
swift-module-digester: diagnose value ownership changes for parameters.

swift-ci committed:
Merge pull request #19728 from nkcsgexi/diag-ownership-change

swift-ci committed:
Merge pull request #19722 from compnerd/winsk

slavapestov committed:
Merge pull request #19723 from slavapestov/irgen-type-lowering-cleanup

IRGen type lowering cleanup

DougGregor committed:
Merge pull request #19719 from DougGregor/superclass-mangled-name

 [ABI] Use mangled superclass names from class context descriptors.

swift-ci committed:
Merge pull request #19726 from adrian-prantl/43439465

swift-ci committed:
Merge pull request #19730 from rudkx/move-another-fast-test

lorentey committed:
[test] Fix spurious failure in optimized tests

We don’t emit the trap message in optimized builds, so don’t check for it.

lorentey committed:
[test] Confess Set/Dictionary ABI changes

mikeash committed:
Merge pull request #19614 from mikeash/no-internal-export

[Stdlib] Change SWIFT_RUNTIME_STDLIB_INTERNAL to not export the symbol.

swift-ci committed:
Merge pull request #19734 from lorentey/fix-optimized-crash-test

compnerd committed:
Merge pull request #19475 from drodriguez/android-aarch64-adb_push_built_products

[android] adb_push_build_products.py support for aarch64.

Azoy committed:
Update Mangling Docs

eeckstein committed:
Merge pull request #19714 from eeckstein/fix-cowopt

COWArrayOpts: make the optimization work again for two-dimensional arrays.

nkcsgexi committed:
swift-module-digester: re-enable test for Json serialization diagnostics.

DougGregor committed:
[Type checker] Ensure that synthesized accessors get generic parameter lists.

When we synthesize the declaration for getters and setters from (e.g.) an
addressor, ensure that we clone the generic parameter list when we have a
generic subscript. Fixes rdar://problem/45046969.

compnerd committed:
stdlib: use "C" style casts in the overlay shims

When this header is used to create a module, it is used in a "C" environment,
which means that we cannot use the C++ style casts.  Switch to the
old-style-cast.  If we need to, we can always use `-Wno-old-style-casts` to
silence the warnings in C++ mode.

compnerd committed:
stdlib: use `__swift_ssize_t` instead of `ssize_t` (NFC)

`ssize_t` is not available on Windows.  We have an internal type definition
`__swift_ssize_t` which allows us to use the type.  Switch to that spelling.

compnerd committed:
stdlib: include "Windows.h" before "Bcrypt.h"

The system header depends on Windows.h but does not include it itself.  This
results in base Windows types (e.g. ULONG) to be undefined.  Include the header
to include the needed typedefs.

compnerd committed:
stdlib: correct case for the bcrypt import library

On Windows the filesystem is not case sensitive and this will link just fine.
However, the Windows SDK provides the import library with the lowercase name.
Adjust the name so that the link actually succeeds on case-sensitive file
systems (like on Linux).  This fixes the Windows cross-compile.

jrose-apple committed:
Merge pull request #19727 from jrose-apple/emit-parseable-module-interface

[Driver] Expose -emit-parseable-module-interface[-path]

swift-ci committed:
Merge pull request #19736 from Azoy/mangling-docs

slavapestov committed:
Merge pull request #19707 from slavapestov/read-legacy-layout

Add support for reading YAML legacy type info dump

swift-ci committed:
Merge pull request #19738 from DougGregor/synthesize-getter-setter-generic-subscript

slavapestov committed:
Add test case for https://bugs.swift.org/browse/SR-1571

nkcsgexi committed:
swift-module-digester: keep track of whether a property/subscript has didSet and willSet.

swift-ci committed:
Merge pull request #19741 from nkcsgexi/property-observer-module-digester

compnerd committed:
Merge pull request #19739 from compnerd/windows-stdlib

Windows stdlib

Azoy committed:
address Nate's feedback

jrose-apple committed:
Merge pull request #19522 from jrose-apple/lined-pockets

Remove unwanted exceptions from @inlinable checking

nkcsgexi committed:
swift-module-digester: include non-resilient class decls when checking ABI stability.

nkcsgexi committed:
test: regenerate abi stability baseline after recent fixes.

Commit:
Remove the CoreImage overlay as none of the things it contains are public.

compnerd committed:
build: workaround cross-compilation for Windows

When cross-compiling the Windows standard library in release mode, we
would get undefined symbols due to LLVM invoking
`add_definitions(-D_DEBUG)`.  This would change the behaviour of certain
functions from MSVCRT resulting in undefined symbols.  Scrub out the
unwanted `-D_DEBUG`.

swift-ci committed:
Merge pull request #19383 from Azoy/se-0202-changelog

compnerd committed:
stdlib: use some clever C11 to define ssize_t

`ssize_t` is not universally available as it is not a standard type.
However, we require clang to build the runtime due to custom calling
convention support.  As a result, we know that the compiler will support
the `_Generic` keyword from the C11 standard.  Use this with an
expansion to map the type appropriately for all targets.  This avoids
having to have a sanity check in the runtime that the type definition
matches the underlying system.

Commit:
update the modules in the CMakeLists files

slavapestov committed:
AST: Remove Swift 3 access control checking

slavapestov committed:
Parse: Remove Swift 3 support

lorentey committed:
[test] Set and Dictionary now has an extra inhabitant (32-bit)

rudkx committed:
Merge pull request #19749 from rudkx/fix-short-circuiting

[ConstraintSystem] Tweak -disable-constraint-solver-performance-hacks.

slavapestov committed:
Merge pull request #19731 from slavapestov/stray-commits

Couple of stray fixes I forgot to check in

harlanhaskins committed:
[InterfaceGen] Print property initializers in resilient, fixed-layout types (#19619)

Augment the ASTPrinter to print the name and text of initializer expressions if
a property has an initializer and the type is @_fixed_layout and resides in a resilient module, and serialize the text for partial modules.

With this change, all .swiftinterface files in the project (except for SwiftLang) compile to swiftmodules on macOS.

rdar://43774580
rdar://43812188

swift-ci committed:
Merge pull request #19747 from compnerd/defaway

Commit:
Extend operator decls to allow any designated nominal type for lookup.

Rather than limiting this to protocols, allow any nominal type.

Rename -enable-operator-designated-protocols to
-enable-operator-designated-types to reflect the change.

harlanhaskins committed:
[SILOptimizer] Don't diagnose infinite recursion if a branch terminates (#19724)

This patch augments the infinite recursion checker to not warn if a
branch terminates, but still warns if a branch calls into something with
`@_semantics("arc.programtermination_point")`. This way, calling `fatalError`
doesn't disqualify you for the diagnostic, but calling `exit` does.

This also removes the warning workaround in the standard library, and
annotates the internal _assertionFailure functions as
`programtermination_point`s, so they get this treatment too.

slavapestov committed:
Sema: Remove most of the remaining Swift 3 support

Commit:
[ConstraintSystem] Fix mistake involving comparison to extension.

We need to compare the Decl for the nominal we're extending, not the
extension decl itself.

swift-ci committed:
Merge pull request #19753 from rudkx/fix-extension-decl-comparison

swift-ci committed:
Merge pull request #19533 from lorentey/per-instance-seeding

rintaro committed:
[CodeCompletion] Handle TupleShuffleExpr in completion context analyzer

(and TupleExpr at non-call argument position).
Now, unresolved member completion in array literal should work.

Also, Don't calculate convertibility to 'Any' type. That would be a
noise to type relation because anything is convertible to 'Any'.

rdar://problem/43302814

DougGregor committed:
[AST] Ill-formed @objc protocols might have associated types.

As an optimization, we don't even look for associated types in @objc
protocols. However, this could lead to broken invariants like "every
associated type has a witness" in ill-formed @objc protocols that do
have associated types.

Implement this optimization by checking whether the protocol has a
Clang node. Fixes rdar://problem/41425828 / SR-8094.

CodaFi committed:
Drop an unused variable from parameter explosion mapping

CodaFi committed:
Merge pull request #19755 from CodaFi/controlled-expansion

[NFC] Drop an unused variable from parameter explosion mapping

compnerd committed:
stdlib: always build the shims subdirectory

This directory is part of the compiler image.  It installs the shims
needed to build content with the compiler.  Hoist this out of the stdlib
only build to allow installing the SwiftShims module as part of the
compiler.

DougGregor committed:
[AST] Don’t deserialize members of @objc protocol looking for associated types

Thanks to @slavapestov for pointing this out! The optimization to avoid
looking through the members of a protocol that can’t possibly have any
associated types was for both deserialized protocols and imported
protocols, but I only supported the former in my previous change. Check
both cases and make the reasons much more obvious.

Also, that change resolved an existing compiler crasher as well.

DougGregor committed:
[Test] Add test for SR-8666 / rdar://problem/43893945.

This test case crashed at runtime in Swift 4.2, but is now fixed. Add it
so that we don't regress.

swift-ci committed:
Merge pull request #19754 from DougGregor/objc-assoc-type-crash

DougGregor committed:
[Type checker] Check for recursion when evaluating @objc on a property.

Fixes crash from rdar://problem/33093935.

swift-ci committed:
Merge pull request #19758 from DougGregor/objc-var-recursion-crash

compnerd committed:
build: allow specifying the Linux SDK path

The SDK is a sysroot.  Permit the user to specify the location of the
SDK when building Linux.  This is in preparation to cross-compile the
Linux standard library.

While we are exposing more control knobs to the user, we can start
simplifying `configure_sdk_unix`.

compnerd committed:
build: hard code the triple into the SDK configuration

Rather than computing the triple first and passing it down to the SDK
configuration, hard code the triple into the SDK's configuration.  This
is a value which is static and will allow us to configure
cross-compilation of the standard library for Linux.  Although most
other targets do not support the same variety of architectures, this
enables those to be cross-compiled as well.

compnerd committed:
build: remove unnecessary parameter to configure_sdk_unix

Remove the duplicated platform name parameter to the unix SDK
configuration..  This value is implicit currently and can be computed.
This needs to be computed to permit parallel co-installations of
architectures for targets which do not support fat binaries (i.e.
non-MachO targets).

compnerd committed:
build: remove sdk parameters which can be computed

The prefix and platform names are case adjusted values of the name of
the SDK.  Compute the value rather than pass them down.  The
simplifications will yield an easier cross-compilation implementation.

compnerd committed:
build: remove unneeded variables for non-Darwin SDKs

This alters the SDK configuration to remove the variables that don't
impact the target configuration.  Simply don't define them.

compnerd committed:
build: prevent printing for unused variables for SDK

Reduce some of the SDK configuration printing for non-Darwin
configurations.  This simplifies the output and SDK configuration.  The
parameter reduction simplifies the setup for cross-compilation.

compnerd committed:
build: use some variables for the defaults

Use variables for the defaults to normalise the paths through the
various hosts.  This makes a follow up refactoring more obvious.

tokorom committed:
[vim] Add swiftCaselLabelRegion

tokorom committed:
[vim] Add swiftDefaultLabelRegion

graydon committed:
Revert "[SILOptimizer] Don't diagnose infinite recursion if a branch terminates (#19724)"

This reverts commit e94450e840009180e1e5ed90bd89112499bc0927.

rdar://45080912

lorentey committed:
Merge pull request #19688 from lorentey/hashed-bridgeobject

[stdlib] Set, Dictionary: Replace _Variant enums with _BridgeStorage

Commit:
Fix up naming in some places that I missed.

compnerd committed:
test: use llvm-strings instead of strings

We should have the LLVM build which provides most of the tools that we need.
`strings` is amongst them.  Add a new `llvm-strings` substitution from lit and
use that instead of `strings`.  This helps with Windows which does not have a
`strings` tool.

DougGregor committed:
[Type checker] Only "use" bridging conformances when there is a type checker.

Works around rdar://problem/45047761 and rdar://problem/45058722.

Commit:
Do not restore the precedencegroup name when deserializing them.

We only need to have the identifier during type checking of operator
declarations, so we do not need to restore it from the
PrecedenceGroupDecl during deserialization. We can just use the
deserialized name from the PrecedenceGroupDecl directly if needed.

This does result in one change in behavior. When printing modules, we
previously didn't print 'DefaultPrecedence' for items that had no
precedence specified, but now we will as seen in the test update for
IDE/print_ast_tc_decls.swift.

rudkx committed:
Merge pull request #19756 from rudkx/extend-operator-designated-type

Extend operator decls to allow any designated nominal type for lookup.

nkcsgexi committed:
Merge remote-tracking branch 'apple/master' into include-non-resilient-class

swift-ci committed:
Merge pull request #19768 from DougGregor/bridge-conformance-only-when-tc

compnerd committed:
Merge pull request #19757 from compnerd/shims

stdlib: always build the shims subdirectory

compnerd committed:
stdlib: correct declaration for `malloc_usable_size` on android

`malloc_usable_size` is provided on Android with API Level 17 or newer.
However, the signature between bionic and glibc is slightly different, with
android marking the parameter as `const`.  This results in a type mismatch error
when building the standard library for android.  Adjust it to match the target
system.

swift-ci committed:
Merge pull request #19764 from tokorom/vim-syntax-case-label-region

nkcsgexi committed:
swift-module-digester: diagnose non-final function changes in non-resilient classes.

nkcsgexi committed:
swift-module-digester: diagnose adding/removing final as ABI/API breakages.

compnerd committed:
build: fix compatibility with VS2017 and VS2015

Adjust the build to work with the VS2017 layout while maintaining the
compatibility with VS2015.  The library location changed which results in the
link failing due to not finding system libraries.

nkcsgexi committed:
swift-module-digester: avoid including private protocol conformances when checking API stability.

compnerd committed:
Merge pull request #19761 from compnerd/sdk-options

SDK options

nkcsgexi committed:
swift-module-digester: exclude fixed binary order when checking API stability.

swift-ci committed:
Merge pull request #19748 from compnerd/generic-size-types

nkcsgexi committed:
test: regenerate ABI/API stability baselines after recent fixes.

nkcsgexi committed:
Merge pull request #19744 from nkcsgexi/include-non-resilient-class

swift-module-digester: include non-resilient class decls when checking ABI stability.

Commit:
[IDE] Add test case for already fixed crash in SourceEntityWalker

Test case for rdar://problem/45039252 that was already fixed by
a8136a29ba45d0045aefb0a1130d0f286bd532db

compnerd committed:
Merge pull request #19770 from compnerd/bionic-malloc

stdlib: correct declaration for `malloc_usable_size` on android

compnerd committed:
build: remove unnecessary sets (NFC)

The default values are not referenced after the cached value is set above.  The
set only complicates the SDK configuration and accomplishes nothing.  Remove the
dead code.  NFC.

compnerd committed:
build: remove set of internal CMake variable

CMAKE_EXECUTABLE_FORMAT is set by CMake based on the target.  We were setting it
in certain cases.  Uniformly rely on CMake configuring the parameter correctly.
This should have no impact on the build.

swift-ci committed:
Merge pull request #19773 from compnerd/vscompat

compnerd committed:
build: fix the variable name

The parameter name was not changed, resulting in the string conversion
not working properly for the Windows builds.

airspeedswift committed:
Switch UnsafeMutableBufferPointer.subscript to use _modify (#19771)

Commit:
Merge pull request #19776 from nathawes/indexing-crash-test-case

[IDE] Add test case for already fixed crash in SourceEntityWalker

slavapestov committed:
SIL: Correctly substitute type parameters in opened existential types

This wasn't possible when the SIL cloner was first written, but subclass
existentials can involve generic classes, so you can have an opened
archetype type with a superclass constraint involving a type constraint.

harlanhaskins committed:
[Sema] Allow non-public static property initializers in @_fixed_layout types

Static properties are not subject to the same restrictions as properties
whose initializers are exposed to clients in @_fixed_layout types.

rdar://45080912

slavapestov committed:
Merge pull request #19783 from slavapestov/fix-arnold-corner-case

SIL: Correctly substitute type parameters in opened existential types

compnerd committed:
Merge pull request #19780 from compnerd/objformat

build: remove set of internal CMake variable

swift-ci committed:
Merge pull request #19785 from harlanhaskins/static-shock

compnerd committed:
Merge pull request #19767 from compnerd/strings

test: use llvm-strings instead of strings

swift-ci committed:
Merge pull request #19779 from compnerd/names

airspeedswift committed:
Switch .first to use an index not an iterator (#18075)

slavapestov committed:
Merge pull request #19750 from slavapestov/remove-swift-3

Remove some more Swift 3 compatibility hacks

swift-ci committed:
Merge pull request #19778 from compnerd/android-variant

DougGregor committed:
[AST] Imported classes have no stored properties.

Avoid triggering the import of all of the members of an Objective-C
class to determine its stored properties, because we don't import
instance variables of Objective-C classes. rdar://problem/45060773

nkcsgexi committed:
Merge pull request #19737 from nkcsgexi/enable-api-json

swift-module-digester: reenable test for Json serialization diagnostics.

swift-ci committed:
Merge pull request #19789 from DougGregor/no-storage-for-old-classes

rudkx committed:
Merge pull request #19787 from rudkx/extend-operator-designated-type

Do not restore the precedencegroup name when deserializing them.

harlanhaskins committed:
[SILOptimizer] Don't diagnose infinite recursion if a branch terminates the program (#19781)

* [SILOptimizer] Don't diagnose infinite recursion if a branch terminates the program

This patch augments the infinite recursion checker to not warn if a
branch terminates, but still warns if a branch calls into something with
@_semantics("programtermination_point"). This way, calling fatalError
doesn't disqualify you for the diagnostic, but calling exit does.

This also removes the warning workaround in the standard library, and
annotates the internal _assertionFailure functions as
programtermination_points, so they get this treatment too.

* Fix formatting in SILInstructions.cpp

* Re-add missing test

benlangmuir committed:
Merge pull request #19792 from rintaro/ide-completion-rdar43302814

[CodeCompletion] Handle TupleShuffleExpr in completion context analyzer

compnerd committed:
SourceKit: the swift-lang requires the standard library

The swift-lang swift module requires the standard library to be built and is
part of the standard library component.  Ensure that this is taken into
consideration when we attempt to build it.

nkcsgexi committed:
swift-module-digester: add a flag to include Swift decls specifically.

This flag allows us to diagnose the overlay part exclusively without digesting
the entire SDK framework.

nkcsgexi committed:
Merge pull request #19799 from nkcsgexi/module-digester-swift-only

swift-module-digester: add a flag to include Swift decls specifically.

swift-ci committed:
Merge pull request #19798 from compnerd/source-kit-overlay

Commit:
Merge pull request #19745 from lancep/RemoveCoreImageOverlay

Remove the CoreImage overlay as none of the things it contains are marked public

bjhomer committed:
Run tests for the new 'try?' behavior with '-swift-version 5'
